test/storage-luatest/storage_1_1_test.lua

```diff
@@ -261,3 +261,87 @@ test_group.test_on_bucket_event = function(g)
         box.space.data2:drop()
     end)
 end
+
+--
+-- gh-434: bucket_send() shouldn't change the transfer flags if a transfer for
+-- the same bucket is already in progress.
+--
+test_group.test_bucket_double_send = function(g)
+    g.replica_2_a:exec(function()
+        ivshard.storage.internal.errinj.ERRINJ_LAST_RECEIVE_DELAY = true
+    end)
+    local bid = g.replica_1_a:exec(function(uuid)
+        local transfer_flags =
+            ivshard.storage.internal.rebalancer_transfering_buckets
+        local bid = _G.get_first_bucket()
+        local f = ifiber.create(ivshard.storage.bucket_send, bid, uuid,
+                                {timeout = 100000})
+        f:set_joinable(true)
+        rawset(_G, 'test_f', f)
+        ilt.assert(transfer_flags[bid])
+        local ok, err = ivshard.storage.bucket_send(bid, uuid)
+        ilt.assert_equals(err.code, iverror.code.WRONG_BUCKET)
+        ilt.assert(not ok)
+        -- Before the bug was fixed, the second send would clear the flag, thus
+        -- leaving the first sending unprotected.
+        ilt.assert(transfer_flags[bid])
+        return bid
+    end, {g.replica_2_a:replicaset_uuid()})
+    g.replica_2_a:exec(function()
+        ivshard.storage.internal.errinj.ERRINJ_LAST_RECEIVE_DELAY = false
+    end)
+    g.replica_1_a:exec(function(bid)
+        local transfer_flags =
+            ivshard.storage.internal.rebalancer_transfering_buckets
+        local f = _G.test_f
+        _G.test_f = nil
+        local f_ok, ok, err = f:join(iwait_timeout)
+        ilt.assert_equals(err, nil)
+        ilt.assert(f_ok)
+        ilt.assert(ok)
+        ilt.assert(not transfer_flags[bid])
+        _G.bucket_gc_wait()
+    end, {bid})
+    --
+    -- Cleanup.
+    --
+    g.replica_2_a:exec(function(bid, uuid)
+        local ok, err = ivshard.storage.bucket_send(bid, uuid,
+                                                    {timeout = iwait_timeout})
+        ilt.assert_equals(err, nil)
+        ilt.assert(ok)
+        _G.bucket_gc_wait()
+    end, {bid, g.replica_1_a:replicaset_uuid()})
+end
+
+--
+-- gh-434: bucket_recv() shouldn't change the transfer flags if a transfer for
+-- the same bucket is already in progress.
+--
+test_group.test_bucket_double_recv = function(g)
+    g.replica_2_a:exec(function(bid, uuid)
+        local transfer_flags =
+            ivshard.storage.internal.rebalancer_transfering_buckets
+        local f = ifiber.create(ivshard.storage.bucket_recv, bid, uuid, {})
+        f:set_joinable(true)
```

Comment:
Happens rarely:
```
[002] storage-luatest/storage_1_1_test.lua
[002] TAP version 13
[002] 1..2
[002] # Started on Tue Oct 31 21:11:03 2023
[002] # Starting group: storage
[002] ok     1	storage.test_bucket_double_send
[002] not ok 2	storage.test_bucket_double_recv
[002] #   ...ing/tnt/vshard/test/storage-luatest/storage_1_1_test.lua:326: the fiber is dead
[002] #   stack traceback:
[002] #   	...ing/tnt/vshard/test/storage-luatest/storage_1_1_test.lua:322: in function 'storage.test_bucket_double_recv'
[002] #   	...
[002] #   	[C]: in function 'xpcall'
[002] # Ran 2 tests in 0.327 seconds, 1 success, 1 error
```

Reply:
I couldn't reproduce it. It either passes or fails earlier on one of the bugs which I've fixed in another PR. But I did apply this:
```Diff
+++ b/test/storage-luatest/storage_1_1_test.lua
@@ -341,6 +341,7 @@ test_group.test_bucket_double_recv = function(g)
         -- Cleanup.
         --
         _G.bucket_recovery_wait()
+        _G.bucket_gc_wait()
         ilt.assert_equals(box.space._bucket:get{bid}, nil)
     end, {vtest.storage_first_bucket(g.replica_1_a),
           g.replica_1_a:replicaset_uuid()})
```
The recovery turns the bucket into GARBAGE, and it sometimes affects runs of newer groups with other configs.